### ISSUE TITLE
Feature: Callback for target file disk path while creating by MediaUploader

### ISSUE THREAD

**Issue by hasnatbabur** [plank/laravel-mediable#163]
Hi, thanks for this plugin. Can you please provide a callback for setting/changing target disk path dynamically. 

Ideally it will be nice to upload different mimetype files to different disk path automatically. I am not sure if it is safe/can use before_save callback for changing the $directory property of the uploader.

Something like:
```
MediaUploader::fromSource($x)->setTargetPathUsing(function($mediaFile){
    // we have $mediaFile here for detecting mime type
    return 'images/';  // logic where we have the resolved source file object
});
```
Thanks

**Comment by frasmage** [comment]
Hi @hasnatbabur,

You should be able to use the `beforeSave()` callback to do that kind of logic. In case you weren't aware, you can use the aggregate_type definitions in the config to define groupings of mimes, e.g. images, audio, video, documents, etc.

```php
MediaUploader::fromSource($source)
    ->toDisk($disk)
    ->beforeSave(function(Media $media) {
        $media->directory = $media->aggregate_type;
    })->upload();
```

That said, now that I look at it there is a slight bug in that the duplicate file validation and handling happens before that callback, so if you change the directory it won't validate the new destination. I will put up a fix soon to invert the ordering there.

**Comment by frasmage** [comment]
Hey @hasnatbabur, I forgot to reply to this, but the ordering of the callback was fixed in version 4.1.0. You should be able to use beforeSave() as I described above.
